### ISSUE TITLE
instance name :"clock" conflict

### ISSUE THREAD

**Issue by flowmeter** [IowaDave/RTC-DS3231-Arduino-Interrupt#1]
NorthernWidget changed "clock" to myRTC in order to stop conflict with clock instances.

Nethertheless I didn't managed to make example work, and can't find why. There's still a problem between code / library / hardware

_EDIT : It finally worked after having run DS3231_test updated with myRTC instead of clock. There must be an instruction in DS3231_test that make things work._

Anyway this is the best commented code that i've found so far on DS3231. Many thanks

**Comment by IowaDave** [comment]
Firstly, thank you for kind words and for looking at this repository.

It sounds like you found a way to get an example sketch to work. That is good progress. But you are probably working on some project of your own. Is there any code you would like to share here for review?

It does puzzle me to think that a name chosen for an instance of the DS3231 library object would affect the behavior of a sketch. Using "clock" as the instance name? Why not? It could just as easily be "myRTC" or "xyzzy"; the name need only to be unique within the project.

I took a fresh look at NorthernWidget's github repo.  The variable name, "clock" was changed to "myRTC" in the "set_echo" example sketch in late 2021. However, interestingly, as of early 2022 it remained "clock" in the other examples. My own example in this repo uses "clock" too.

Hmm, my code example targeted an AVR-type of board, such as an Uno or a Leonardo. Northern Widget's note mentioned "cleaning up" the set_echo sketch for use with SAMD microprocessors. Ignorant, old me wonders whether the name, "clock" appears in any of the core libraries for SAMD-based Arduino boards? In that case, then perhaps it would be best to avoid using "clock" as a name for an object or variable in a user sketch. Were you compiling for a SAMD-based board, such as a Zero, Nano33, or one of the MKR series?

Happy to take a look at anything you care to share. My best to you.

**Comment by flowmeter** [comment]
Hello Dave,
I had hard time making this library work, especially with interrupt.
i just had an "ahah" moment when re-reading "enable Alarm" I understood that if one wants a falling edge interruption he needs to write in his code both
`checkIfAlarm(1);
checkIfAlarm(2);`

if not, one will never get 5 volts to SQW pin.

Maybe we should give more code examples. I can share my tests codes with you.
Best
And thank you very much 1000 times !

**Comment by IowaDave** [comment]
Many thanks to you, also. I revised the README file to provide more detail about the alarm flag bits.

I'm glad you found your way out of the difficulty. And I'm grateful that you shared the info. This author and other readers benefit from your efforts!

Dave

**Comment by IowaDave** [comment]
Reopening to ask a question. @flowmeter, did you happen to use the "DS3231_set" program that comes with the NorthernWidget DS3231 library? It's just a guess, but I imagine that a lot of people might use it as a convenient way to set the time on their DS3231 after they install a battery.

If you did, then it might be what caused your problem.  It sets both Alarm 1 and Alarm 2. Go back and look at the code for that program. Near the end, around lines 99-109, come the following code statements:

```
		// Test of alarm functions
		// set A1 to one minute past the time we just set the clock
		// on current day of week.
		Clock.setA1Time(DoW, Hour, Minute+1, Second, 0x0, true, 
			false, false);
		// set A2 to two minutes past, on current day of month.
		Clock.setA2Time(Date, Hour, Minute+2, 0x0, false, false, 
			false);
		// Turn on both alarms, with external interrupt
		// Turn on both alarms, with external interrupt
		Clock.turnOnAlarm(1);
		Clock.turnOnAlarm(2);
```
Two minutes after the time you set, Alarm 2 goes off. The Alarm 2 flag gets set. And the SQW pin will stay LOW as long as the Alarm 2 pin stays set.

Could that be what happened to you?

People who want to use the "DS3231_set" example to set the time on their RTC module -- but not set the alarms -- may want to consider deleting, or commenting-out those lines.

Users of the DS3231 RTC device can definitely turn off both of the alarm flags during setup. They can do so with the statements you mentioned in your previous post: 
```
  checkIfAlarm(1); 
  checkIfAlarm(2);
```

My example program (in this repo) uses only Alarm 1. And I wrote my own short sketch to set the clock initially. Which might explain why I never ran into a problem where Alarm 2 was holding the SQW pin LOW. I learned something from this issue. My example will now clear both of the alarm flags during the setup() function, out of an abundance of caution.

**Comment by flowmeter** [comment]
Hi Dave,

your guess is right ! "Reopening to ask a question. @flowmeter, did you happen to use the "DS3231_set" program that comes with the NorthernWidget DS3231 library? "

Here is a code that i used to test functions in order to make my project work. I still have a bug with the sleep function at the first loop. If I clear this bug with a wire it never re-appears after... 

I've made comments for you in the code. I updated my Github profile to ease communication.

ACTION is a switcher i use in a SWITCH / CASE option directions in my operating code with 4 interrupt functions. I simplified it here.

I'm sorry if i made mistakes; programming is not my job, I'm a newbie, and was born in the early 60s. Feel free to change, simplify, ease readability of code. You can release it in examples section. I don't deserve to be quoted.

Best,
Emmanuel

![SerialMonitorDS3231](https://user-images.githubusercontent.com/23551048/158161144-ce5aee78-5596-49a0-be69-897933768879.JPG)

/!\ beware there are many mistakes in this code, but i'm fixing them by edits. 2022/05/12.

`// RTC DS3231 functions test
//
// for DS3231 functions see : https://github.com/NorthernWidget/DS3231/blob/master/keywords.txt
//
#include <DS3231.h> // andrew wickert V1.1.0
#include <Wire.h>
const unsigned long BAUD_RATE = 115200; // choose yours

    // Direction switcher
volatile byte ACTION = 60;

    // debouncing setup
const byte Delay_without_bounce = 10; // ms (no more, if not RTC digitalinterrupt won't wake up board from deep sleep !)

//  need to clean variables names someday
DS3231 myRTC;

byte year;
byte month;
byte Date;
byte DoW;
byte hour;
byte minute;
byte second;

byte ADay, AHour, AMinute, ASecond, A1Bits , A2Bits; // 
bool ADy, A12h, Apm;

// controls
const byte frequency = 1; // alarm clock frequency (one every (frequency) minutes)

void setup() {
// To trigger alarms you just need to wait for RTC alarm clock (pin D2),
// To manually wake up board you just need a simple wire between ground and D3 pin
   pinMode(2, INPUT_PULLUP);
    attachInterrupt(digitalPinToInterrupt(2), digitalInterruptCASE1, FALLING);
   pinMode(3, INPUT_PULLUP);
    attachInterrupt(digitalPinToInterrupt(3), digitalInterruptCASE1, FALLING); 
    
    Serial.begin(BAUD_RATE);
    Wire.begin();
    myRTC.enable32kHz(HIGH); // @Dave please check if necessary and comment why

// uncoment functions that you want to test during set-up
    // Alrm1setup();
    // Alrm2setup();
    // myRTC.turnOffAlarm(1);
    // myRTC.turnOffAlarm(2);
    // myRTC.turnOnAlarm(1);
    // myRTC.turnOnAlarm(2);
    /*if (myRTC.checkIfAlarm(1))
    {
      Serial.println("Alarm *1* triggered");
    }*/
    Showwatch();
    Reprogramalrm1();
   ShowAlrm1hour();
   delay(1000);
   myRTC.turnOffAlarm(1);
   myRTC.checkIfAlarm(1);
   myRTC.turnOnAlarm(1);
    myRTC.turnOffAlarm(2);
   myRTC.checkIfAlarm(2);
}

void loop() {
// uncoment functions that you want to test during loop   
   // Alrm1setup();
   // Showtemp(); 
   // runTheSpecialCode(); // current time first method
   // Showwatch();  // current time another method
   // Showalrm1status();
   // myRTC.turnOnAlarm(1);
   // Showalrm1status();
   // ShowAlrm1hour();
   // ShowAlrm2hour();
   // Showalrmbits(); 
   // Reprogramalrm1(); // Automaticaly sets next alarm at (frequency) minutes
   // Reprogramalrm2();
   //
   delay(10);
   if (ACTION == 1)
   {
    Serial.println("I woke-up at ");
    Showwatch();
    delay(100);
    Serial.println("Next alarm at ");
    Reprogramalrm1();
    delay(1000);
    ACTION = 60;
    }
   else 
   {
    Serial.println("I'm going to sleep ");
    delay(1000);
    SleepR3();
   }
    
    /*if (myRTC.checkIfAlarm(1))
    {
      Serial.println("Alarm 1 triggered");
      Reprogramalrm1();
      ShowAlrm1hour();
    }*/
} // loop end

void SleepR3(){  // taken from here https://www.youtube.com/watch?v=urLSDi7SD8M
// To put Arduino to sleep and wait for an interrupt. See : kdcircuits.com
//Disable ADC - don't forget to flip back after waking up if using ADC in your application ADCSRA |= (1 << 7);
//
  ADCSRA &= ~(1 << 7);
    //ENABLE SLEEP - this enables the sleep mode langage assembleur
  SMCR |= (1 << 2); //power down mode
  SMCR |= 1;//enable sleep
 // __asm__ __volatile__("sleep"); !!! Uncomment when used without BOD disable !!!



  //BOD DISABLE - this must be called right before the __asm__ sleep instruction
  MCUCR |= (3 << 5); //set both BODS and BODSE at the same time
  MCUCR = (MCUCR & ~(1 << 5)) | (1 << 6); //then set the BODS bit and clear the BODSE bit at the same time
  __asm__  __volatile__("sleep");//in line assembler to go to sleep

  // in order to come back from deep sleep you need a digital input LOW on interrupt pins
} // SleepR3() end

void digitalInterruptCASE1(){ 
  static u32 precedenteIT = millis();
  u32 _millis = millis();
  if ((_millis - precedenteIT) >= Delay_without_bounce) // /!\  beware of Delay_without_bounce using a RTC
  {
   ACTION = 1;
  }
  precedenteIT = _millis;
}

void Alrm1setup() // set up your alarm time
{
  myRTC.setA1Time(
      7, // alarmTime.day(),
      12, // alarmTime.hour(),
      02, // alarmTime.minute(),
      00, // alarmTime.second(),
      0x00001110, // this mask means alarm when seconds match
      false, false, false
    );
    Serial.println(" Alarm 1 set");
} //Alrm1setup end

void Alrm2setup()
{
  myRTC.setA2Time(
      1, // alarmTime.day(),
      10, // alarmTime.hour(),
      21, // alarmTime.minute(),
      0x0001100, // this mask means alarm when hours match
      false, false, false
    );
    Serial.println(" Alarm 2 set");
} // Alrm2setup end

void Showtemp()
{
   Serial.println();
   Serial.print("Temperature = ");
   Serial.print(myRTC.getTemperature(),2);
   Serial.println(" °C");
} // Showtemp end

void Showwatch()
{
  DateTime now = RTClib::now();    // not a good idea to use same name now for variable and method, shame on me
      delay(10);
      Serial.println();
      Serial.println("Right now it's :");
      
      Serial.print(now.day());
      Serial.println(" Day");

      Serial.print(now.month());
      Serial.println(" Month");
      
      Serial.print(now.year());
      Serial.println(" Year");
            
      Serial.print(myRTC.getDoW()); // /!\ method is changed in order to work /!\ //
      Serial.println(" Day of the week");
      
      Serial.print(now.hour());
      Serial.println(" Hour");
      
      Serial.print(now.minute()); // the current minute, 0..59
      Serial.println(" Minutes");
      
      Serial.print(now.second());
      Serial.println(" Seconds");
      Serial.println();
} // Showwatch end

void Showalrm1status()
{
  boolean test1 = 0;
  test1 = myRTC.checkAlarmEnabled(1);
  if (test1 != 1) {
    Serial.println(" A1 disabled");
    Serial.print("test = ");
    Serial.println(test1);
    }
    else {
    Serial.println(" A1 activated");
    Serial.print("test = ");
    Serial.println(test1);
    }
} // Showalrm1status end

void Showalrm2status()
{
  boolean test2 = 0;
  test2 = myRTC.checkAlarmEnabled(2);
  if (test2 != 1) {
    Serial.println(" A2 disabled");
    Serial.print("test = ");
    Serial.println(test2);
    }
    else {
    Serial.println(" A2 activated");
    Serial.print("test = ");
    Serial.println(test2);
   }
} // Showalrm2status end

void ShowAlrm1hour()
{
  delay(10);
  Serial.print("Alarm 1 : ");
   myRTC.getA1Time(ADay, AHour, AMinute, ASecond, A1Bits, ADy, A12h, Apm);
   Serial.print(ADay, DEC);
   if (ADy) {
      Serial.print("th day of the week");
   } else {
      Serial.print("th day of the month");
   }
   Serial.print(" ");
   Serial.print(AHour, DEC);
   Serial.print("h ");
   Serial.print(AMinute, DEC);
   Serial.print("mn ");
   Serial.print(ASecond, DEC);
   Serial.println("sec ");
    if (A12h) {
      if (Apm) {
      Serial.print(" pm");
      } else {
      Serial.print( "am");
      }
    }
    if (myRTC.checkAlarmEnabled(1)) {
      Serial.println(" Activated");
    }
} // ShowAlrm1hour end

void ShowAlrm2hour()
{
  Serial.print("Alarme 2 : ");
   myRTC.getA2Time(ADay, AHour, AMinute, A2Bits, ADy, A12h, Apm);
   Serial.print(ADay, DEC);
   if (ADy) {
      Serial.print(" th day of the week");
   } else {
      Serial.print(" th day of the month");
   }
   Serial.print(" ");
   Serial.print(AHour, DEC);
   Serial.print(" : ");
   Serial.print(AMinute, DEC);
   if (A12h) {
      if (Apm) {
      Serial.print(" pm");
      } else {
      Serial.print( "am");
      }
    }
    if (myRTC.checkAlarmEnabled(2)) {
      Serial.println(" Activated");
    }
} // ShowAlrm2hour end

void Showalrmbits()
{
  // Display alarm bits
    Serial.print("Alarme 1 Mask bits : ");
    Serial.println(A1Bits, BIN); 
    Serial.print("Alarme 2 Mask bits : ");
    Serial.println(A2Bits, BIN);
} // Showalrmbits end

void Reprogramalrm1() // Sets next alarm time by adding (frequency) minutes to current time
{
DateTime now = RTClib::now(); // Get current RTC time
byte Alminute =0;
byte Alhour =0;
byte Alday =0;
        Date = now.day();
        hour = now.hour();
        minute = now.minute();
        second = now.second();
        if (minute + frequency >=60) // ...if to wake at >= 60 minutes       
        {
        Alhour = hour + 1; // sets next hour
        Alminute = frequency + minute - 60; // computes minutes modulo 60
          if (hour == 23) // ... if we're close to midnight
          {
            Alday = Date + 1; // sets next day // /!\ beware it doesn't work at the end of the months and year, will edit this further
            Alhour =0; // ...at 0 hour
          }
          else
          {
            Alday = Date;
          }
        }
         else 
          {
          Alminute = frequency + minute;
          Alday = Date;
          Alhour = hour;
          }
        Serial.println("Next Alarm1 is set to : ");  
        Serial.print("date = ");
        Serial.println(Alday);
        Serial.print("hour = ");
        Serial.println(Alhour);
        Serial.print("minute = ");
        Serial.println(Alminute);
        Serial.print("seconds = ");
        Serial.println(second);
        myRTC.setA1Time(Alday, Alhour, Alminute, second, 0x0, false,false, false);
        myRTC.turnOffAlarm(2);
        myRTC.checkIfAlarm(2);
        myRTC.turnOffAlarm(1);
        myRTC.checkIfAlarm(1);
        myRTC.turnOnAlarm(1);
} // Reprogramalrm1 end

void Reprogramalrm2() // Sets next alarm time by adding (frequency) minutes to current time
{
DateTime now = RTClib::now(); // Get current RTC time
byte Alminute =0;
byte Alhour =0;
byte Alday =0;
        Date = now.day();
        hour = now.hour();
        minute = now.minute();
        if (minute + frequency >=60) // ...if to wake at >= 60 minutes       
        {
        Alhour = hour + 1; // sets next hour
        Alminute = frequency + minute - 60; // computes minutes modulo 60
          if (hour == 23) // ... if we're close to midnight
          {
            Alday = Date + 1; // sets next day
            Alhour =0; // ...at 0 hour
          }
          else
          {
            Alday = Date;
          }
        }
         else 
          {
          Alminute = frequency + minute;
          Alday = Date;
          Alhour = hour;
          }
        Serial.println("Next Alarm2 is set to : ");  
        Serial.print("date = ");
        Serial.println(Alday);
        Serial.print("hour = ");
        Serial.println(Alhour);
        Serial.print("minute = ");
        Serial.println(Alminute);
        myRTC.setA2Time(Alday, Alhour, Alminute, 0x0, false, false,false);
} // Reprogramalrm2 end

void runTheSpecialCode() {
  // get the current time
  // using the DateTime and RTClib classes
  // defined in DS3231.h
  DateTime dt = RTClib::now();

  // print the current time
  Serial.print("Year ");
  Serial.print(dt.year()); Serial.print(":");
  Serial.print("Month ");
  Serial.print(dt.month()); Serial.print(":");
  Serial.print("Day ");
  Serial.print(dt.day()); Serial.print(":");
  Serial.print(dt.hour()); Serial.print(":");
  Serial.print("Hour ");
  if (dt.minute() < 10) Serial.print("0");
  Serial.print(dt.minute()); Serial.print(":");
  Serial.print("Minutes ");
  if (dt.second() < 10) Serial.print("0");
  Serial.print(dt.second()); 
  Serial.println("Seconds "); 
} // runTheSpecialCode end`

**Comment by flowmeter** [comment]
Just to share a very good page on DS3231
https://lastminuteengineers.com/ds3231-rtc-arduino-tutorial/

**Comment by IowaDave** [comment]
Thanks! Yes, that's a good article, as far as it goes. The alarms and the
interrupt get only passing mention, unfortunately.

I'm not familiar with the Adafruit RTC library. The work I wrote about here
made use of the Wickert (Northern Widget) library. There are yet more,
other, different libraries too.  As someone once said, "Yer takes yer pick
and then yer takes yer chances!"

Donor-provided libraries are nice, and welcome, and mostly useful. They can
also be a source of unexpected behavior due to code one did not write and
might not fully understand.

The DS3231 just looks like external memory to an Arduino. A person could
dispense with an RTC library and go direct. Tom Almy details how to
approach it in the "Two Wire Interface AKA I2C" chapter of his terrific
book, "Far Inside the Arduino."

**Comment by flowmeter** [comment]
I focused only on hardware explanations, not on adafruit library.

Arduino is an awesome machine, if one can learn with a sensei. Right now i'm throwing shit code lines to it because i didn't get the right lessons from the start and can't find someone bringing me by the hand.

I'm trying to build an iot scale working on solar NRJ, sending messages with a GSM shield, without sending bucks to trash. I had to shift from UNO to pro mini to lower power consumption. It's a walk on the wild side for me.

**Comment by IowaDave** [comment]
You are not alone. I know how you feel. Have often felt that way myself.
Keep going! Each thing you learn is like a nugget of gold that you can put
in your pocket and keep forever.

Some books that have helped me on the hardware side of things include:


   - *Far Inside the Arduino* by Tom Almy
   - *Arduino Software Internals* by Norman Dunbar


All I can do here is to hope that what I learned and shared about the
DS3231 has been of some benefit to you. Please consider that other people
might be interested in things you learn as you make progress.

When you get your project working, I hope you'll consider sharing what you
learned with others by writing it up here on github!

**Comment by flowmeter** [comment]
> When you get your project working, I hope you'll consider sharing what you
learned with others by writing it up here on github!

That's what i began to do, with the code shared MArch 14th.
I edit it some times because i'm working on it, and i'm fixing my bugs and errors.

Thanks for the book's references.

**Comment by IowaDave** [comment]
I understand now that some misunderstanding may exist. Allow me to clear it
up, as follows:

1. I did look at your code dated March 14, 2022.
2. I did not respond to it because I did not understand that any
expectation attached to me that I would provide feedback.
3. Today I looked at it again, and as a courtesy I am prepared to provide
the feedback given below.
4. The feedback below will bring my involvement in your project to an end.
I'm simply not in a position to help you with your code.

You mentioned that you have continued to work on the code, which means the
code of March 14 may have changed in some ways. I cannot possibly comment
on any changes to the code, and as mentioned above I will not be in a
position to look at it again. Therefore, the comments I feel able to offer
must be limited to the code of March 14.

Surely by now you have dealt with the syntax issues. Out of an abundance of
respect I won't mention those.

I don't think you want the 32KHz signal to appear on the SQW pin of the
DS3231. Try leaving that turned off. I believe you want to use the SQW pin
for its other purpose, which is to signal interrupts when an alarm goes off.

I notice that the same interrupt handler gets attached to two, different
external interrupt pins: both pin 2 and pin 3. It is not clear which of the
pins carries the interrupt signal from the DS3231. As to the other pin, if
it is to be activated by some other, different device, then perhaps it
needs a different interrupt handler.

A general comment about interrupt handlers: keep them as short as possible
One reason is that interrupts are disabled globally while an interrupt
service routine (ISR) is running. Your code contemplates receiving two,
different interrupts. You probably don't want to miss a second one while
processing the first. Perhaps only set the signal value inside the ISR,
e.g., SIGNAL = 1. Let the main event loop detect the signal value and take
appropriate actions there.

A comment specifically on the interrupt handler in the March 14 code. It
looks like it tries to wait for some time to pass before setting the SIGNAL
value. But I wonder whether it will work.

If the ISR has been giving you problems, try implementing the wait outside
of the ISR.

Finally, I see so many different functions that it's hard to comprehend
what the program is trying to accomplish. Consider breaking the problem
down into smaller pieces, then take the pieces one at a time. I'm sure you
have tried what I'm going to suggest next, but perhaps if you try it again
you'll see something that might have escaped notice previously.

I'll write this in pseudo-code, and defer to you for the actual code
writing. My suggestion is to get this (and the other functions your code
needs to perform) all working properly before you attack the challenges of
power-down sleep mode.

#include... maybe list Wire.h first, then DS3231, since DS3231 uses the
Wire library.

define a "sentinel" variable, such as "ACTION" in the current code, for the
ISR to notify the main event that an interrupt was received

setup () {
  start the communications
  set the time on the DS3231 if necessary
  set and enable an alarm for, say, a minute later
  be sure to clear the alarm flags in the DS3231, too
  attach the ISR to the pin that goes to the SQW pin of the DS3231
  set the sentinel variable to a value that means "false"
}

main event loop {
   check the value of the sentinel variable
   if it is "true" then take appropriate actions {
      clear the alarm flag in the DS3231
      set the sentinel variable to a value that means "false"
      set another alarm in the DS3231, if you wish
      print something via Serial to exhibit details of the interrupt
   }
}

ISR {
  set the sentinel variable to a value that means "true"
}

Get that working reliably first. Later, you can add other functions to it.

About power-down sleep: you're trying to save power? Power-down sleep can
significantly reduce the power consumption of an AVR processor running
alone, but not so much when it is mounted on an Arduino development board.
The reason is that the Arduino hardware consumes power too, and it
continues to do so even when the AVR is asleep.

You can find really good, detailed, helpful explanations of power-saving
sleep, and of interrupts in the two books by Tom Almy and Norman Dunbar,
mentioned previously.

I have run out of things to say about the March 14 code, and I cannot
undertake to remain engaged with you on this project in the future. I
encourage you to press on. I feel confident that you will figure it out,
probably sooner than you think!

Finally, again, please accept my gratitude for your interest in my article
about DS3231.

And now, I must bow out of the picture.

**Comment by flowmeter** [comment]
Hello Dave,
Thank you for your long comments. I really appreciate your involvement in helping others.

Sorry for my bad communication. It was not for help, although I had some incertainties on some parts (32kHz, arduino not wakening - because of my anti-bouncing time, why now.Dow() is not working while myRTC.getDoW() is, ...)

I just posted this piece of code to put all examples together and share an easy way to test functions about alarms in setup or loop. This by just uncommenting functions calls.
This is why you don't understand what is the purpose of the code. The code has no goal outside being a test platform.

One can use the deep sleep function in order to test DS3231 wakening arduino. Interrupt function is coded both for DS3231 automatic interrupt and manual interrupt with a simple wire one can plug in Gnd pin, or via a push button. Anti-bouncing is to make sure only one interrupt is triggered.

Yes power management is power hungry and low power projects use bare µprocessors connected to battery. So I switched to pro mini 3.3V. I'm still at POC stage using out of he shelf devices.

All your posts made me understand how alarms work or why they don't work.
Thank you very much for what you shared.
